fancyimpute/missforest.py

```diff
@@ -0,0 +1,249 @@
+        self.imp_continuous_index = None
+        self.imp_categorical_index = None
+
+        Solver.__init__(self,
+            min_value=min_value,
+            max_value=max_value)
+
+    def solve(self, X, missing_mask):
+        X = check_array(X, force_all_finite=False)
+        self.sorted_col = self.sort_col(missing_mask)
+        self.coltype_dict = self._judge_type(X)
+
+        self.imp_continuous_index, self.imp_categorical_index = \
+                self.get_type_index(missing_mask, self.coltype_dict)
+
+        differ_categorical = float('inf')
+        differ_continuous = float('inf')
+
+        init_fill = X
+
+        while self.stop is False:
+
+            differ_categorical_old = differ_categorical
+            differ_continuous_old = differ_continuous
+
+            x_old_imp = init_fill
+
+            x_new_imp = []
+
+            for col in self.sorted_col:
+                tmp = []
+                if self.coltype_dict[col] is 'categorical':
+                    model = self.rf_cla
+                else:
+                    model = self.rf_reg
+
+                x_obs, y_obs, x_mis = self.split(init_fill, col, missing_mask)
+                model.fit(x_obs, y_obs)
+                y_mis = model.predict(x_mis)
+                for ele in y_mis:
+                    tmp.append(ele)
+                    x_new_imp.append(ele)
+                init_fill[:, col][missing_mask[:,col]] = tmp
+            x_new_imp = np.asarray(x_new_imp)
+
+            differ_continuous, differ_categorical = self._lose_func(x_new_imp, x_old_imp)
+            if differ_continuous >= differ_continuous_old and differ_categorical >= differ_categorical_old:
+                self.stop = True
+        return init_fill
+
+    def _lose_func(self, imp_new, imp_old):
+        """
+        Evaluation Method, mathematical concept are available at 'https://www.stu-zhouyc.com/iterForest/metrics'
+
+        :param imputed_data_old: a dict like {'col name':[predicted value1,...],...}
+                                        the dict contains original missing index which is part of the original data
+                                        its the last estimated data
+                                        accompany with brand-new imputed data, they are going to be evaluate.
+        :return:
+        """
+
+        continuous_imp_new = imp_new[self.imp_continuous_index]
+        continuous_imp_old = imp_old[self.imp_continuous_index]
+        categorical_imp_new = imp_new[self.imp_categorical_index]
+        categorical_imp_old = imp_old[self.imp_categorical_index]
+
+        try:
+            continuous_div = continuous_imp_new - continuous_imp_old
+            continuous_div = continuous_div.dot(continuous_div)
+            continuous_sum = continuous_imp_new.dot(continuous_imp_new)
+
+            categorical_count = np.sum(categorical_imp_new == categorical_imp_old)
+            categorical_var_len = len(categorical_imp_new)
+
+        except:

```

Comment:
What kind of errors do you expect here? Can you be more explicit in catching them?

Reply:
Yes, its moved, I forget this, my bad